File: Algorithms/103.BinaryTreeZigzagLevelOrderTraversal.py
class Solution:
    # @param {TreeNode} root
    # @return {integer[][]}
    def zigzagLevelOrder(self, root):
        result = []
        if not root: return result
        queue, next_queue, values, flag = [root], [], [], True
        while queue:
            node = queue.pop(0)
            values.append(node.val)
            if node.left: next_queue.append(node.left)
            if node.right: next_queue.append(node.right)
            if not queue:
                if flag:
                    result.append(values[:])
                else:
                    result.append(values[::-1])
                values, queue, next_queue, flag = [], next_queue, [], not flag
        return result

File: Algorithms/java/103.BinaryTreeZigzagLevelOrderTraversal.java
public class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        if (root == null) {
            return result;
        }
        List<Integer> v = new ArrayList<>();
        List<TreeNode> stack = new ArrayList<>();
        List<TreeNode> next_level_stack = new ArrayList<>();
        stack.add(root);
        boolean flag = false;
        while (stack.size() != 0) {
            TreeNode p = stack.get(0);
            stack.remove(0);
            if (flag) {
                v.add(0, p.val);
            }
            else {
                v.add(p.val);
            }
            if (p.left != null) {
                next_level_stack.add(p.left);
            }
            if (p.right != null) {
                next_level_stack.add(p.right);
            }
            if (stack.size() == 0) {
                stack = next_level_stack;
                next_level_stack = new ArrayList<>();
                result.add(v);
                v = new ArrayList<>();
                flag = !flag;
            }
        }
        return result;
    }
}
